gpflow/base.py

```diff
@@ -60,18 +69,26 @@ def __init__(self,
             self._unconstrained = tf.Variable(unconstrained_value,
                                               dtype=dtype, name=name, trainable=trainable)
 
-        self.prior = prior
-
     def log_prior(self):
-        x = self.read_value()
-        y = self._unconstrained
+        """ Prior probability density.
+        This can be evaluated either on the constrained or unconstrained variable.
+        For example if we set transform = Exp(), prior = Uniform(), prior_on_constrained=True
+        Then log_prior will either be uniform (when evaluate_on_constrained is True),
+        or log_prior scales as 1/value when evaluate_on_constrained is set to False.
+        """
 
         if self.prior is not None:
-            out = tf.reduce_sum(self.prior.log_prob(x))
-            if self.transform is not None:
-                log_det_jacobian = self.transform.forward_log_det_jacobian(y, y.shape.ndims)
-                out += tf.reduce_sum(log_det_jacobian)
-            return out
+            y = self.read_value()
+            if self.prior_on == PriorOn.CONSTRAINED:
+                log_p = tf.reduce_sum(self.prior.log_prob(y))
+            else:
+                x = self._unconstrained
+                log_p = tf.reduce_sum(self.prior.log_prob(x))
+                if self.transform is not None:
+                    log_det_jacobian = self.transform.inverse_log_det_jacobian(y, y.shape.ndims)
+                    log_p += tf.reduce_sum(log_det_jacobian)
```

Comment:
I thought this branch (prior on unconstrained) was the default behaviour before - but this code has x and y swapped compared to the old version?

Reply:
Right, the x and y have swapped definitions to maintain consistency with how they are defined in tensorflow bijectors. (Is this what you meant?) It is also a bit more intuitive if we think of eg y=exp(x) for the Exp() bijector